link-header-parser.lisp

```diff
@@ -0,0 +1,189 @@
+  (let ((results (make-link-header)))
+    (link-record data results)
+    (link-header-records results)))
+
+(defun signal-error (expected-token data)
+  (error "expected ~a at the start of ~s" expected-token data))
+
+(defun consume-blanks (data)
+  (multiple-value-bind (token rest-data)
+      (next-token data)
+    (if (and token
+             (token= token :blanks))
+        rest-data
+        data)))
+
+(defun field-separator (data)
+  (multiple-value-bind (token rest-data)
+      (next-token data)
+    (if (and token
+             (token= token :field-separator))
+        (values t rest-data)
+        nil)))
+
+(defun record-separator (data)
+  (multiple-value-bind (token rest-data)
+      (next-token data)
+    (if (and token
+             (token= token :record-separator))
+        (values t rest-data)
+        nil)))
+
+(defun link-record (data results)
+  (let* ((record-results (make-link-record))
+         (url-rest-data  (url data record-results))
+         (no-blanks-rest-data (consume-blanks url-rest-data)))
+    (multiple-value-bind (field-separator-found field-separator-rest-data)
+        (field-separator no-blanks-rest-data)
+      (let ((next-record no-blanks-rest-data))
+        (if field-separator-found
+          (let* ((no-blanks-rest-data  (consume-blanks field-separator-rest-data))
+                 (parameters-rest-data (parameters no-blanks-rest-data record-results)))
+            (push record-results (link-header-records results))
+            (setf next-record parameters-rest-data))
+          (push record-results (link-header-records results)))
+        (let* ((rest-after-blanks (consume-blanks next-record)))
+          (multiple-value-bind (record-separator-found record-separator-rest-data)
+              (record-separator rest-after-blanks)
+            (if record-separator-found
+                (let ((new-records (consume-blanks record-separator-rest-data)))
+                  (link-record new-records results))
+                url-rest-data)))))))
+
+(a:define-constant +url-wrapper-chars+ '(#\< #\>) :test #'equalp)
+
+(defun url (data record-results)
+  (multiple-value-bind (token no-url-data)
+      (next-token data)
+    (if (not (and token
+                  (token= token :url)))
+        (signal-error :url no-url-data)
+        (progn
+          (setf (link-record-url record-results)
+                (string-trim +url-wrapper-chars+ (token-value token)))
+          no-url-data))))
+
+(defun consume-parameter-separator (data)
+  (multiple-value-bind (token rest-data)
+      (next-token data)
+    (if (and token
+             (token= token :parameter-separator))
+        rest-data
+        (signal-error :parameter-separator data))))
+
+(a:define-constant +parameter-value-wrapper-char+ '(#\") :test #'equalp)
+
+(defun parameters (data parameters-results)
+  (multiple-value-bind (token rest-data)
+      (next-token data)
+    (let ((key   nil)
+          (value nil))
+      (if (not (and token
```

Comment:
Please never use `(if .. (progn ..) ..)`, `cond` is a much more readable choice.

Reply:
Sure, no problem!

I hope i fixed every place where this problem occurred, now.